### PR TITLE
joint training of composition model

### PR DESCRIPTION
**Pull request by Blubberli**
This PR contains the functionality to jointly train a composition model on two tasks. The model takes as input two words and constructs three representations. for example:
word1: hot
word2: weather

the vectors for "hot" and "temperature" are transformed by a number of matrices and result in the same number of transformed vectors
representation 1: separate weighting combines the transformed vectors into representation 1: this representation is trained to be close to the original phrase (gold(hot_ weather)) 
representation 2: separate weighting combines the transformed vectors into representation 2: this representation is trained to be close to the implicit relation /attribute that holds between hot in weather which is temperature
composed: representation 1 and 2 are combined into a final representation by weighting each and adding them (and then normalizing them). Thus, this representation should capture the general meaning of the phrase but also the implicit attribute

we compute two losses:
loss 1: cosine distance(representation 1 , original phrase representation)
loss 2: cosine distance(representation 2, attribute)
we back propagate based on the combined loss (loss 1 + loss 2). 
--> those could be weighted but when I ran the model the losses for both tasks were more or less balanced (the one for phrase reconstruction was a bit lower)

I added one epoch of pretraining because the reconstruction task theoretically has a lot more different labels to lean and thus,  is more complicated. This could be optional or even left out if there is not a big difference but I would do that in a second PR

I currently evaluate the model on attribute ranking based on the nearest neighbor ranker. I report the result for representation 2 (attribute specific representation) as well as for the final representation. It might make sense to add representation 1 as well later on.

To train the model I needed to add a new dataset type but it is very simple. it takes two datasets and returns a batch for each, so that when you iterate in training you get two batches at the same time, one contains phrases with attributes and one phrases with phrase representations. I compute loss one for the phrase batch and loss 2 for the attribute batch and use these losses to backpropagate

a test is added as well.

the PR is very long, especially the training script, so take your time to review it. 

- we probably want to add a simpler model (matrix) model as a joint model as well. but we can do that in a second PR....

### CODE REVIEW

File: training_scripts/joint_training.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+    :param predictions_att_rep: the predicted representations (the attribute-specific representation)
+    :param ranks_final_phrase: the path to save the ranks to (for final phrase)
+    :param ranks_att_rep: the path to save the ranks to (for the attribute)
+    :param dataset: the dataset to evaluate on
+    :param labels: all possible labels that can be predicted
+    :param config: the configuration that corresponds to the data
+    """
+    rank_loader = torch.utils.data.DataLoader(dataset, batch_size=len(dataset), num_workers=0,
+                                              shuffle=False)
+    ranker_attribute = NearestNeigbourRanker(path_to_predictions=predictions_att_rep,
```
Comment (evvahu) [inline training_scripts/joint_training.py:201]: needs to be changed according to changes in NearestNeighbourRanker

File: training_scripts/joint_training.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+    :return: predictions and losses for the learned attribute and the final composed representation, the original phrases
+    """
+    test_loss_att = []
+    test_loss_final = []
+    predictions_final_rep = []
+    predictions_attribute_rep = []
+    orig_phrases = []
+    model.to(device)
+    for batch in test_loader:
+        batch["device"] = device
```
Comment (evvahu) [inline training_scripts/joint_training.py:150]: I don't really understand these lines (batch["device"] = device). why do you save device in the config dict?

Comment (evvahu) [reply training_scripts/joint_training.py:150]: okay, I see it's because "model" takes the batch object as its argument.

File: training_scripts/joint_training.py [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+        composed, rep1, rep2 = model(batch)
+        composed = composed.squeeze().to("cpu")
+        rep2 = rep2.squeeze().to("cpu")
+        for pred in rep2:
+            predictions_attribute_rep.append(pred.detach().numpy())
+        for pred in composed:
+            predictions_final_rep.append(pred.detach().numpy())
```
Comment (evvahu) [inline training_scripts/joint_training.py:157]: do I get this right that you have to call .detach because you can't call numpy() on a variable that requires grad?

**Review by evvahu — CHANGES_REQUESTED** [review]
I only request one change in evaluate (training scripts), the other comments are questions for my personal understanding. 

I have some other questions/remarks that I'd like to ask here:
- for this model we will have to tune: dropout, nr of transformations and weights for combining the two representations right? 
- it'd be nice if we had a simpler, for example, matrix model to compare it to. I'd like to take on this task if that's okay? I would like to programme something again so that I can practice more.
-it would also be nice if we can parameterise the pretraining as you mentioned. that shouldn't be difficult.
-as you said we should possibly also evaluate on the representation of the phrase. In the end we should probably use both representations (the composed one and the phrase only one) to do WSD? I'm not sure whether you've already tried it out (I think not?)

and now a really stupid question I should have asked ages ago because it also applies to the ranking models in general. It might have been clear to me at some point but at the moment "stehe ich völlig auf dem Schlauch".
So in the case of the present model, you calculate cosine_distance_loss based on the predicted representation (e.g. for 'hot weather' and the attribute which looks hopefully somewhat like 'temperature') and the gold representation (both the phrase and the attribute). Where do we get the gold representations from?  

I have been a bit absent in terms of the programming tasks in this project and forgotten some stuff already so I'd like to engage more if time allows.

File: training_scripts/joint_training.py (hunk 2 above)
Comment (Blubberli) [reply training_scripts/joint_training.py:150]: it doesn't make a difference when you train on cpu but on gpu, you need to explicitly set the device to gpu at the positions were you want the operations to be performed.

File: training_scripts/joint_training.py (hunk 3 above)
Comment (Blubberli) [reply training_scripts/joint_training.py:157]: yes, we only want the array and not the variable. but I think there are other options to that. e.g we can call pred.data to only get the array and np.array(pred.data) to do the same thing in general (I think, haven't tested that code).

**Comment by Blubberli** [comment]
> I only request one change in evaluate (training scripts), the other comments are questions for my personal understanding.
> 
> I have some other questions/remarks that I'd like to ask here:
> 
> * for this model we will have to tune: dropout, nr of transformations and weights for combining the two representations right?
> * it'd be nice if we had a simpler, for example, matrix model to compare it to. I'd like to take on this task if that's okay? I would like to programme something again so that I can practice more.
>   -it would also be nice if we can parameterise the pretraining as you mentioned. that shouldn't be difficult.
>   -as you said we should possibly also evaluate on the representation of the phrase. In the end we should probably use both representations (the composed one and the phrase only one) to do WSD? I'm not sure whether you've already tried it out (I think not?)
> 
> and now a really stupid question I should have asked ages ago because it also applies to the ranking models in general. It might have been clear to me at some point but at the moment "stehe ich völlig auf dem Schlauch".
> So in the case of the present model, you calculate cosine_distance_loss based on the predicted representation (e.g. for 'hot weather' and the attribute which looks hopefully somewhat like 'temperature') and the gold representation (both the phrase and the attribute). Where do we get the gold representations from?
> 
> I have been a bit absent in terms of the programming tasks in this project and forgotten some stuff already so I'd like to engage more if time allows.

Thanks for the extensive feedback and remarks. 
1) Yes you are right with the parameters
2) simpler model would be nice. matrix model sounds good, though we are facing one problem: the matrix model only has one parameter block (one matrix). thus, if we jointly train this model there are no task specific parameters. one possibility would be to try it without task specific parameters, but then you could also just smash the two datasets into one, increase the number of labels and that's it. another option would be to keep one matrix as a joint parameter but add one task-specific matrix on top of the composed representation for each task. so after retrieving the general composed representation (that should capture the similarities between both tasks) we can transform it with a 200x200 matrix for an attribute representation and with a different 200x200 matrix for the phrase representation. this model would still be simpler and smaller than the transformation model but would make sense in terms of multitask learning. what do you think? I'd be happy if you could program something like that.
3) we should evaluate the distinct representations on WSD. We can even evaluate all three representations (phrase, attribute, combined)
4) the gold representation for the attribute is taken from the same embedding space as all other Word Embeddings (in this case temperature). of course by doing that we make a strong assumption, namely that attributes have a meaningful representation in this vector space. that's why I always need to rephrase attributes like Art&Weise because I cannot get an embedding for that. previous work on attribute learning used other types of representations (i.e. a specific attribute vector space, and then you really have each dimension corresponding to a specific word in context. Hartung has done that before word embedding became a thing and then in the 2017 paper compared these to Word Embeddings (he used w2vec and also did not have a representation for each of his attributes) and you can see that the "traditional" representations are not very powerful unfortunately. But nevertheless, I don't think that every attribute is perfectly represented by a pretrained word embedding and other representations might be worth a try (though I don't see time to do that at the moment)

ok, hoffe ich hab jetzt nix vergessen^^

File: training_scripts/joint_training.py (hunk 1 above)
Comment (Blubberli) [reply training_scripts/joint_training.py:201]: how do I need to change that?

Comment (Blubberli) [reply training_scripts/joint_training.py:201]: done
